### PR TITLE
refactor(label): convert files to typescript FE-4861

### PR DESCRIPTION
**Pull request by edleeks87**
### Proposed behaviour

Convert `Label` component files to TS

### Current behaviour

`Label` component is JS

### Checklist

- [x] Commits follow our style guide
- [x] Unit tests added or updated if required
- [x] Typescript `d.ts` file added or updated if required

#### QA

- [ ] Tested in CodeSandbox/storybook
- [ ] Add new Cypress test coverage if required
- [ ] Carbon implementation matches Design System/designs
- [ ] UI Tests GitHub check reviewed if required

### Testing instructions

Test for regressions in inputs that use label

The following CodeSandbox is an example of the broken behaviour.
You can see the new behaviour by looking at the version in the comment by `codesandbox[bot]`.

https://codesandbox.io/s/carbon-quickstart-typescript-forked-7xpo14?file=/src/index.tsx
Will likely need to export the project locally and confirm no types are broken

### CODE REVIEW

File: src/__internal__/label/label.component.tsx [hunk 1]
```diff
@@ -1,25 +1,73 @@
...
 
-const tooltipPosition = ({ error, warning, info, inline }) => {
-  return (error || warning || info) && inline === true ? "top" : "right";
+const tooltipPosition = ({
+  error,
+  warning,
+  info,
+  inline,
+}: ValidationProps & { inline?: boolean }) => {
+  return (error || warning || info) && inline ? "top" : "right";
```
Comment (grabkowski) [inline src/__internal__/label/label.component.tsx:63]: I think that the return type should be set as `TooltipPositions` type imported from `tooltip.config.ts`

File: src/__internal__/label/label.style.ts [hunk 2]
```diff
@@ -1,7 +1,7 @@
 import styled, { css } from "styled-components";
-import PropTypes from "prop-types";
+import { LabelProps } from "./label.component";
 
-const LabelStyle = styled.label`
+const LabelStyle = styled.label<Pick<LabelProps, "isRequired" | "disabled">>`
```
Comment (grabkowski) [inline src/__internal__/label/label.style.ts:4]: I think that now will be the best time change `LabelStyle` to `StyledLabel` to match our naming convention, what do you think about that?

Comment (grabkowski) [inline src/__internal__/label/label.style.ts:4]: what do you think about defining `StyledLabelProps` here and then extending the component interface with it?

File: src/__internal__/label/label.style.ts [hunk 3]
```diff
@@ -24,12 +24,11 @@ const LabelStyle = styled.label`
...
-LabelStyle.propTypes = {
-  disabled: PropTypes.bool,
-  isRequired: PropTypes.bool,
-};
+const DEFAULT_CONTAINER_WIDTH = 30;
 
-export const StyledLabelContainer = styled.div`
+export const StyledLabelContainer = styled.div<
+  Pick<LabelProps, "align" | "inline" | "pr" | "pl" | "width" | "optional">
+>`
```
Comment (grabkowski) [inline src/__internal__/label/label.style.ts:31]: what do you think about defining `StyledLabelContainerProps` here and then extending the component interface with it?

File: src/__internal__/label/label.style.ts (hunk 2 above)
Comment (edleeks87) [reply src/__internal__/label/label.style.ts:4]: I prefer the approach you're suggest but I'd like this approach to be adopted in the other components we're refactoring as well just so the codebase is consistent etc. I'll raise this with the rest of the team in slack and make the changes if we get agreement

File: src/__internal__/label/label.style.ts (hunk 3 above)
Comment (edleeks87) [reply src/__internal__/label/label.style.ts:31]: See above

**Review by DipperTheDan — COMMENTED** [review]
Nothing more to add from me. Just for you to address the comments from @grabkowski

**Review by DipperTheDan — DISMISSED** [review]

**Review by grabkowski — APPROVED** [review]

**Review by divyajindel — DISMISSED** [review]

**Review by DipperTheDan — APPROVED** [review]

**Review by divyajindel — APPROVED** [review]

**Pull request merged**